src/Simulation/Simulators/SparseSimulator/Native/quantum_state.hpp

```diff
@@ -0,0 +1,1177 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <random>
+#include <cmath>
+#include <functional>
+#include <algorithm>
+#include <list>
+#include <iostream>
+
+#include "basic_quantum_state.hpp"
+
+#include "types.h"
+#include "gates.h"
+
+using namespace std::literals::complex_literals;
+
+namespace Microsoft::Quantum::SPARSESIMULATOR
+{
+
+// power of square root of -1
+inline amplitude iExp(int power)
+{
+    int p = ((power % 4) + 8) % 4;
+    switch (p)
+    {
+    case 0:
+        return 1;
+    case 1:
+        return 1i;
+    case 2:
+        return -1;
+    case 3:
+        return -1i;
+    default:
+        return 0;
+    }
+    return 0;
+}
+
+template<size_t num_qubits>
+bool get_parity(std::bitset<num_qubits> bitstring){
+    return bitstring.count() % 2;
+}
+
+// Compares two bitsets as through they were bitstrings
+// Used to enforce an ordering on bitsets, though currently not referenced
```

Comment:
If this is currently not used, it can be removed completely. If it is used and critical for performance, we should optimize this implementation (it currently uses many bitset operations on the full width).

Reply:
It is used in a loop where we need to swap entries. We swap only half the time :)